### DISCUSSION TITLE
Environment variable file ordering

### DISCUSSION THREAD

**Discussion by erkde** [vercel/next.js#16530] [Help] [answered]
So, I'm new to NextJS so still getting my head around how some things work.

But whenever I run dev, or build, within a next project, I notice it prints out the loading order of the .env files, for example [in this sample project](https://github.com/erkde/nextjs-blog), I get:

```
% yarn dev   
yarn run v1.22.4
$ next dev
[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
> Info: Loaded env from .env.development.local
> Info: Loaded env from .env.development
> Info: Loaded env from .env
[ ready ] compiled successfully - ready on http://localhost:3000
```

This order is surprising to me, I would expect it to be the other way - to first load .env with the values common for all environments, and then load the more specific files after, overriding more general values.

When I run that sample project, it defines the same environment NEXT_PUBLIC_ENV_VARIABLE in each of the three files, and the value that wins is from the last file, .env.

I see the commit that added [this line](https://github.com/vercel/next.js/blame/canary/packages/next/lib/load-env-config.ts#L82) mention fixing the dotenv loading order to match the [dotenv "spec"](https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use), which looks to me to say the priority of specific files is higher, that .env has the lowest priority.

Or is this intentionally like this in NextJS?

**Comment by lfades** [comment, marked as answer]
@erkde Environment variables won't be replaced. i.e `.env.development.local` has a higher priority than `.env`.

For a more in depth explanation of how it works check out the official example: https://github.com/vercel/next.js/tree/canary/examples/environment-variables

**Reply by erkde** [reply]
> @erkde Environment variables won't be replaced. i.e `.env.development.local` has a higher priority than `.env`.
> 
> For a more in depth explanation of how it works check out the official example: https://github.com/vercel/next.js/tree/canary/examples/environment-variables

yeah, I have read that - hmm, I must be doing something really wrong - in my sample app, I have three environment files, containing the same value, e.g.

```
.env.development.local
NEXT_PUBLIC_ENV_VARIABLE="I'm from .env.development.local"

.env.development
NEXT_PUBLIC_ENV_VARIABLE="I'm from .env.development"

.env
NEXT_PUBLIC_ENV_VARIABLE="I'm from .env"
```

and in `pages/index.jsx`, I just console log the value of that variable...

```
% npm run dev

> learn-starter@0.1.0 dev /Users/erik/Coding/erkde/nextjs-blog
> next dev

info  - Loaded env from /Users/erik/Coding/erkde/nextjs-blog/.env.development.local
info  - Loaded env from /Users/erik/Coding/erkde/nextjs-blog/.env.development
info  - Loaded env from /Users/erik/Coding/erkde/nextjs-blog/.env
ready - started server on http://localhost:3000
event - compiled successfully
event - build page: /
wait  - compiling...
event - compiled successfully
The value of NEXT_PUBLIC_ENV_VARIABLE is I'm from .env
```

If `.env.development.local` did have highest priority, wouldn't I expect to see "I'm from .env.development.local" ?

**Reply by lfades** [reply]
Unfortunately I can't reproduce the behavior you're seeing. the value from `.env.development.local` should be coming first 🤔

**Reply by lfades** [reply]
I created a PR to help with this: https://github.com/vercel/next.js/pull/16536 

We currently don't mention this on docs to not confuse people, but that's okay for the example.

**Reply by erkde** [reply]
Ok, to reproduce, there's a link to a sample app in the initial comment:

```
$ git clone git@github.com:erkde/nextjs-blog.git
$ cd nextjs-blog
$ npm i
$ npm run dev

> learn-starter@0.1.0 dev /Users/erik/Coding/erkde/tmp/nextjs-blog
> next dev

info  - Loaded env from /Users/erik/Coding/erkde/tmp/nextjs-blog/.env.development.local
info  - Loaded env from /Users/erik/Coding/erkde/tmp/nextjs-blog/.env.development
info  - Loaded env from /Users/erik/Coding/erkde/tmp/nextjs-blog/.env
ready - started server on http://localhost:3000
event - compiled successfully
event - build page: /next/dist/pages/_error
wait  - compiling...
event - compiled successfully
event - build page: /
wait  - compiling...
event - compiled successfully
The value of NEXT_PUBLIC_ENV_VARIABLE is I'm from .env
```

Doesn't seem like it works the way your PR describes?

**Reply by blazzy** [reply]
> Unfortunately I can't reproduce the behavior you're seeing. the value from .env.development.local should be coming first

By should be coming first do you mean should be loaded first? That sounds like incorrect behavior. I would expect it to be loaded _last_ so the values in the file would have the highest precedence and be the values actually used.

**Comment by tomas2387** [comment]
It seems this issue was left unanswered. 

Next.js correctly prioritizes env variables from .env, .env.development, .env.local, etc. 
As we can read from the [dotenv docs](https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use):
1. If the current machine already has ENV variables set, those variables won't be replaced. 
2. dotenv assumes the deployed environment has more knowledge about configuration than the application does. 

To debug:      
- Run `env` command to see machine environment variables.    
- If you are using docker-compose, look for a `env_file: .env` in your docker-compose.yml and remove it to let dotenv handle environment variables.

**Comment by lbittner-pdftron** [comment]
+1 the order the variables are loaded seems backwards to me. In a production environment I would expect my `.env.production` to get loaded first.

**Comment by bonesoul** [comment]
having the exact same issue here..

In my case .env.production overwrites anything in .env.production.local where according the documentation (https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order) it shouldn't be so.

Am I hitting a bug or something here?

**Reply by bonesoul** [reply]
my issue was caused by using dotenv package in my common/ package that was used by next.js project. removing the dotenv fixed the ordering issue.

**Reply by sansmoraxz** [reply]
Yeah am facing the same with `.env` and `.env.local`
